net/dns/nrpt_windows.go

```diff
@@ -0,0 +1,297 @@
+	// termination of the Tailscale process, and depending on the
+	// rule, it may prevent us from reaching login.tailscale.com to
+	// boot up. The bootstrap resolver logic will save us, but it
+	// slows down start-up a bunch.
+	ret.DelAllRuleKeys()
+	return ret
+}
+
+func (db *nrptRuleDatabase) loadRuleSubkeyNames() {
+	result := winutil.GetRegStrings(nrptRuleIDValueName, nil)
+	if result == nil {
+		// Use the legacy rule ID if none are specified in our registry key
+		result = []string{nrptSingleRuleID}
+	}
+	db.ruleIDs = result
+}
+
+// initWriteAsGP determines which registry path should be used for writing
+// NRPT rules. If there are rules in the GP path that don't belong to us, then
+// we should use the GP path.
+func (db *nrptRuleDatabase) initWriteAsGP() {
+	var err error
+	defer func() {
+		if err != nil {
+			db.writeAsGP = false
+		}
+	}()
+
+	dnsKey, err := registry.OpenKey(registry.LOCAL_MACHINE, dnsBaseGP, registry.READ)
+	if err != nil {
+		return
+	}
+	defer dnsKey.Close()
+
+	ki, err := dnsKey.Stat()
+	if err != nil {
+		return
+	}
+
+	// If the dnsKey contains any values, then we need to use the GP key.
+	if ki.ValueCount > 0 {
+		db.writeAsGP = true
+		return
+	}
+
+	if ki.SubKeyCount == 0 {
+		// If dnsKey contains no values and no subkeys, then we definitely don't
+		// need to use the GP key.
+		db.writeAsGP = false
+		return
+	}
+
+	// Get a list of all the NRPT rules under the GP subkey.
+	nrptKey, err := registry.OpenKey(registry.LOCAL_MACHINE, nrptBaseGP, registry.READ)
+	if err != nil {
+		return
+	}
+	defer nrptKey.Close()
+
+	gpSubkeyNames, err := nrptKey.ReadSubKeyNames(0)
+	if err != nil {
+		return
+	}
+
+	// Add *all* rules from the GP subkey into a set.
+	gpSubkeyMap := make(map[string]struct{}, len(gpSubkeyNames))
+	for _, gpSubkey := range gpSubkeyNames {
+		gpSubkeyMap[strings.ToUpper(gpSubkey)] = struct{}{}
+	}
+
+	// Remove *our* rules from the set.
+	for _, ourRuleID := range db.ruleIDs {
+		delete(gpSubkeyMap, strings.ToUpper(ourRuleID))
+	}
+
+	// Any leftover rules do not belong to us. When group policy is being used
+	// by something else, we must also use the GP path.
+	db.writeAsGP = len(gpSubkeyMap) > 0
+}
+
+// DelAllRuleKeys removes any and all NRPT rules that are owned by Tailscale.
+func (db *nrptRuleDatabase) DelAllRuleKeys() error {
+	if err := db.delRuleKeys(db.ruleIDs); err != nil {
+		return err
+	}
+	if err := winutil.DeleteRegValue(nrptRuleIDValueName); err != nil {
+		db.logf("Error deleting registry value %q: %v", nrptRuleIDValueName, err)
+		return err
+	}
+	db.ruleIDs = nil
+	return nil
+}
+
+// delRuleKeys removes the NRPT rules specified by nrptRuleIDs from the
+// Windows registry. It attempts to remove the rules from both possible registry
+// keys: the local key and the group policy key.
+func (db *nrptRuleDatabase) delRuleKeys(nrptRuleIDs []string) error {
+	for _, rid := range nrptRuleIDs {
+		keyNameLocal := nrptBaseLocal + `\` + rid
```

Comment:
Is there no equivalent to `filepath.Join()` for registry keys? :-(

I always end up worrying about random quoting type issues, but I suppose the left is a constant and the right is a GUID, so it should be fine.

Reply:
No, there's nothing like that, unfortunately.